File: count_primes.erl
-module(count_primes).
-export([count_primes/1]).

%% 计数质数
%% https://leetcode-cn.com/problems/count-primes/description

-spec count_primes(N :: integer()) -> integer().
count_primes(N) ->
  length(lists:filter(fun(X) -> is_prime(X) end, lists:seq(0, N - 1))).


is_prime(2) -> true;
is_prime(N) when N < 2 orelse N rem 2 == 0 -> false;
is_prime(N) -> is_prime(N, 3).

is_prime(N, K) when N < K * K -> true;
is_prime(N, K) when N rem K == 0 -> false;
is_prime(N, K) -> is_prime(N, K + 2).

File: count_primes.ex
# 计数质数
# https://leetcode-cn.com/problems/count-primes/description

defmodule Solution do
  @spec count_primes(n :: integer) :: integer
  def count_primes(n) do
    Enum.filter(0..(n - 1), &is_prime/1) |> length
  end

  defp is_prime(2), do: true
  defp is_prime(n) when n < 2 or rem(n, 2) == 0, do: false
  defp is_prime(n), do: is_prime(n, 3)

  defp is_prime(n, k) when n < k * k, do: true
  defp is_prime(n, k) when rem(n, k) == 0, do: false
  defp is_prime(n, k), do: is_prime(n, k + 2)
end
